benchmark/scenarios/request_load.py

```diff
@@ -0,0 +1,220 @@
+    requests at a specified rate.
+
+    :ivar reactor: Reactor to use.
+    :ivar request_rate: The target number of requests per second.
+    :ivar sample_size: The number of samples to collect when measuring
+        the rate.
+    :ivar timeout: Maximum time in seconds to wait for the requested
+        rate to be reached.
+    :ivar setup_instance: instance of an implementation of `IScenarioSetup`
+    :ivar request_generator_instance: instance of an implementation of
+        `IRequestGenerator`
+    """
+
+    def __init__(
+        self, reactor, request_generator_instance, request_rate=10,
+        sample_size=DEFAULT_SAMPLE_SIZE, timeout=45,
+        setup_instance=None
+    ):
+        self._maintained = Deferred()
+        self.reactor = reactor
+        self.request_generator = request_generator_instance
+        self.setup = setup_instance
+        self.request_rate = request_rate
+        self.timeout = timeout
+        self.rate_measurer = RateMeasurer(sample_size)
+        self.max_outstanding = 10 * request_rate
+        # Send requests per second
+        self.loop = LoopingCall.withCount(self._request_and_measure)
+        self.loop.clock = self.reactor
+        self.monitor_loop = LoopingCall(self.check_rate)
+        self.monitor_loop.clock = self.reactor
+
+    def _request_and_measure(self, count):
+        """
+        Update the rate with the current value and send `request_rate`
+        number of new requests.
+
+        :param count: The number of seconds passed since the last time
+            `_request_and_measure` was called.
+        """
+        for i in range(count):
+            self.rate_measurer.update_rate()
+
+        def handle_request_error(result):
+            self.rate_measurer.request_failed(result)
+            write_failure(result)
+
+        for i in range(self.request_rate):
+            d = self.request_generator.make_request()
+            self.rate_measurer.request_sent()
+            d.addCallbacks(self.rate_measurer.response_received,
+                           errback=handle_request_error)
+
+    def _fail(self, exception):
+        """
+        Fail the scenario. Stop the monitor loop and throw the
+        error.
+        """
+        self.monitor_loop.stop()
+        self._maintained.errback(exception)
+
+    def check_rate(self):
+        """
+        Verify that the rate hasn't decreased and that the scenario is
+        not overloaded. A scenario is overloaded if there are too many
+        outstanding requests.
+
+        :raise: `RequestRateTooLow` if the rate has dropped.
+        :raise: `RequestOverload` if the scenario is overloaded.
+        """
+        rate = self.rate_measurer.rate()
+        if rate < self.request_rate:
+            self._fail(RequestRateTooLow(rate))
+
+        if self.rate_measurer.outstanding() > self.max_outstanding:
+            self._fail(RequestOverload())
+
+    def start(self):
+        if self.setup is None:
```

Comment:
Rather than using an `if` here, have the caller provide a no-op setup function when nothing needs to be done:
```
def setup():
    return succeed()
```